### PR TITLE
Officially Support Higher-Order Types

### PR DESCRIPTION
**Pull request by LighghtEeloo**
Today, I finally pinned down the long-haunting bug of higher-order types in our type system and fixed it, passing all previously failing tests. I feel ready enough to bring this PR onto the table. The feature enabled by the PR and previous commits is the ability to pass around higher-order types at both type- and term- level, including the type variables in parametrized types, forall types, and exists types.

Before this PR the feature is almost complete, but a bug in finding the least upper bound of two types that are both neutral type applications (in the sense of NbE) occasionally drops the arguments. This PR fixes the troublemaker and provides some tests.

### CODE REVIEW

File: zydeco-lang/src/statics/tyck/lub.rs [hunk 1]
```diff
@@ -112,41 +125,49 @@ impl Lub for Type {
                 let body = lhs.body.inner_clone().lub(rhs.body.inner_clone(), ctx, span)?;
                 Ok(TypeAbs { params: lhs.params, body: rc!(lhs.body.span().make(body)) }.into())
             }
-            (SynType::TypeApp(TypeApp { tvar: NeutralVar::Var(tvar), args: _ }), _)
```
Comment (LighghtEeloo) [inline zydeco-lang/src/statics/tyck/lub.rs:115]: Perhaps the only important thing in this PR: fix how `TypeApp` is dealt with in `Type::lub`.

Comment (ricky136973) [reply zydeco-lang/src/statics/tyck/lub.rs:115]: Does `Type::lub` here stand for `least upper bound`?

**Review by ricky136973 — APPROVED** [review]
Ça semble parfait.

File: zydeco-lang/src/statics/tyck/lub.rs (hunk 1 above)
Comment (LighghtEeloo) [reply zydeco-lang/src/statics/tyck/lub.rs:115]: A quick recap: the least upper bound was introduced to provide the ability to reason about holes in bi-directional type checking. It's an operation that takes two normalized types named `A` and `B` and inductively checks whether a. they have the same structure and b. they only differ on leaves by compatible holes. The operation also returns a more precise type.

The base cases are
```
_ \/ A = A
A \/ _ = A
```

And if neither `A` nor `B`  have `_` (hole) at the top level, check whether the scrutinees are the same and inductively get their children's lub.

**Review by akaiDing — APPROVED** [review]
Looks good to me!

**Pull request merged**